internal/grpctest/tlogger/tlogger.go

```diff
@@ -85,56 +113,111 @@ func log(t *testing.T, format string, fatal bool, args ...interface{}) {
 	logger.t = t
+	logger.errors = nil
+}
+
+// Expect declares an error to be expected. For the next test, all error logs
+// matching the expression (using FindString) will not cause the test to fail.
+// "For the next test" is includes all the time until the next call to Update().
+func Expect(expr string) {
+	re, err := regexp.Compile(expr)
+	if err != nil {
+		logger.t.Error(err)
+		return
+	}
+	logger.errors = append(logger.errors, re)
+}
+
+func (g *tLogger) expected(s string) bool {
+	for _, re := range g.errors {
+		if re.FindStringIndex(s) != nil {
+			return true
+		}
+	}
+	return false
 }
 
 func (g *tLogger) Info(args ...interface{}) {
-	log(g.t, "", false, args...)
+	log(g.t, logLog, "", args...)
 }
 
 func (g *tLogger) Infoln(args ...interface{}) {
-	log(g.t, "", false, args...)
+	log(g.t, logLog, "", args...)
 }
 
 func (g *tLogger) Infof(format string, args ...interface{}) {
-	log(g.t, format, false, args...)
+	log(g.t, logLog, format, args...)
 }
 
 func (g *tLogger) Warning(args ...interface{}) {
-	log(g.t, "", false, args...)
+	log(g.t, logLog, "", args...)
 }
 
 func (g *tLogger) Warningln(args ...interface{}) {
-	log(g.t, "", false, args...)
+	log(g.t, logLog, "", args...)
 }
 
 func (g *tLogger) Warningf(format string, args ...interface{}) {
-	log(g.t, format, false, args...)
+	log(g.t, logLog, format, args...)
 }
 
 func (g *tLogger) Error(args ...interface{}) {
-	log(g.t, "", false, args...)
+	log(g.t, errorLog, "", args...)
 }
 
 func (g *tLogger) Errorln(args ...interface{}) {
-	log(g.t, "", false, args...)
+	log(g.t, errorLog, "", args...)
 }
 
 func (g *tLogger) Errorf(format string, args ...interface{}) {
-	log(g.t, format, false, args...)
+	log(g.t, errorLog, format, args...)
 }
 
 func (g *tLogger) Fatal(args ...interface{}) {
-	log(g.t, "", true, args...)
+	log(g.t, fatalLog, "", args...)
 }
 
 func (g *tLogger) Fatalln(args ...interface{}) {
-	log(g.t, "", true, args...)
+	log(g.t, fatalLog, "", args...)
 }
 
 func (g *tLogger) Fatalf(format string, args ...interface{}) {
-	log(g.t, format, true, args...)
+	log(g.t, fatalLog, format, args...)
 }
 
 func (g *tLogger) V(l int) bool {
 	return l <= g.v
 }
+
+var lcFailed uint32
+
+type errorer struct {
+	t *testing.T
+}
+
+func (e errorer) Errorf(format string, args ...interface{}) {
+	atomic.StoreUint32(&lcFailed, 1)
+	e.t.Errorf(format, args...)
+}
+
+// Tester is an implementation of the x interface parameter to
+// grpctest.RunSubTests with default Setup and Teardown behavior. Setup updates
+// the tlogger and Teardown performs a leak check. Embed in a struct with tests
+// defined to use.
+type Tester struct{}
+
+// Setup updates the tlogger.
+func (Tester) Setup(t *testing.T) {
+	Update(t)
+}
+
+// Teardown performs a leak check.
+func (Tester) Teardown(t *testing.T) {
+	if atomic.LoadUint32(&lcFailed) == 1 {
+		return
+	}
+	leakcheck.Check(errorer{t: t})
+	if atomic.LoadUint32(&lcFailed) == 1 {
+		t.Log("Leak check disabled for future tests")
+	}
+}
```

Comment:
Leak checking isn't really related to the tlogger, so I would prefer to remove it from here.

How about moving this struct into the grpctest package instead?  I like the strategy of embedding it so this doesn't need to be repeated, though.  Nice idea!

Reply:
In addition to leak checking, Tester also does the tlogger update. This would make grpctest dependent on the tlogger, which is something we tried to avoid on the initial tlogger PR.

Reply:
My concern in that PR was about adding default functionality to `grpctest.RunTests` that made it only suitable for grpc.  Having a type we can embed but others could ignore would be fine with me.

I _would_ like to keep the `RunTests` functionality completely separate from grpc, but the "right" way to do that would be to make a sub-package for it specifically.  I don't think that's worth it at this point, but if you do want to move it I can update my references to it from the feature request bug I filed against Go.  But I'm OK with just adding this `Tester` type to it instead.